frame/democracy/src/tests/proxying.rs

```diff
@@ -0,0 +1,104 @@
+// Copyright 2017-2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! The tests for functionality concerning proxying.
+
+use super::*;
+
+#[test]
+fn proxy_should_work() {
+	new_test_ext().execute_with(|| {
+		assert_eq!(Democracy::proxy(10), None);
+		assert!(System::allow_death(&10));
+
+		assert_noop!(Democracy::activate_proxy(Origin::signed(1), 10), Error::<Test>::NotOpen);
+
+		assert_ok!(Democracy::open_proxy(Origin::signed(10), 1));
+		assert!(!System::allow_death(&10));
+		assert_eq!(Democracy::proxy(10), Some(ProxyState::Open(1)));
+
+		assert_noop!(Democracy::activate_proxy(Origin::signed(2), 10), Error::<Test>::WrongOpen);
+		assert_ok!(Democracy::activate_proxy(Origin::signed(1), 10));
+		assert_eq!(Democracy::proxy(10), Some(ProxyState::Active(1)));
+
+		// Can't set when already set.
+		assert_noop!(Democracy::activate_proxy(Origin::signed(2), 10), Error::<Test>::AlreadyProxy);
+
+		// But this works because 11 isn't proxying.
+		assert_ok!(Democracy::open_proxy(Origin::signed(11), 2));
+		assert_ok!(Democracy::activate_proxy(Origin::signed(2), 11));
+		assert_eq!(Democracy::proxy(10), Some(ProxyState::Active(1)));
+		assert_eq!(Democracy::proxy(11), Some(ProxyState::Active(2)));
```

Comment:
What is the advantage to the two way handshake on proxy?

Why can't an account directly give another account proxy, and then leave it to the other account to use or not use their proxy vote?

Reply:
Is it only for the reference counter?

Reply:
because an account can only be a proxy for a single other account and once that account is a proxy it gets a data reference and thus you would be able to mildly grief any target account.